### PR TITLE
Added IgnoreHeuristics message

### PR DESCRIPTION
**Pull request by mohit-a21**
IgnoreHeuristics message can be used to define different
keys on which to ignore tracing of certain calls. For now
the message contains only a url regex. Over time we can
add support for method and host also.

### CODE REVIEW

**Comment by davexroth** [comment]
heuristic is a little generic.  how about ignoreRequests?

File: config.proto [hunk 1]
```diff
@@ -92,3 +95,9 @@ enum PropagationFormat {
     // see https://www.w3.org/TR/trace-context/
     TRACECONTEXT = 1;
 }
+
+// IgnoreHeuristics represents the heuristics used to ignore certain calls from tracing
+message IgnoreHeuristics {
+    // url_regex defines the regular expression for the request path to be ignored
```
Comment (pavolloffay) [inline config.proto:101]: Is the regex evaluated against the full URL or just path part? 

An example of the regex would be helpful.

**Comment by pavolloffay** [comment]
> and host also

If the regex is evaluated against the full URL then the host should be covered by this pattern. 


Maybe I don't remember correctly but @jcchavezs had a use case to ignore tracing when a specific header was present. - e.g. this is another example of ignoring.

File: config.proto [hunk 2]
```diff
@@ -92,3 +95,9 @@ enum PropagationFormat {
     // see https://www.w3.org/TR/trace-context/
     TRACECONTEXT = 1;
 }
+
+// IgnoreHeuristics represents the heuristics used to ignore certain calls from tracing
+message IgnoreHeuristics {
```
Comment (pavolloffay) [inline config.proto:100]: I am not sure I like the `Heuristics` suffix.

**Comment by pavolloffay** [comment]
>  how about ignoreRequests?

I am not sure if I like this either. E.g. what if we will be ignoring DB or messaging spans.

 Maybe `IgnoreMatchers`?

File: config.proto (hunk 1 above)
Comment (mohit-a21) [reply config.proto:101]: The regex is evaluated against the path part only (without query params)
eg: 
```
      - "\/echo\/live"                     # matches /echo/live
      - "\/user\/.*\/cart".                # matches /user/name/cart
      - "\/user\/[0-9]*\/cart".       # matches /user/10/cart
```

**Comment by jcchavezs** [comment]
I think it isn't easy to structure this objects. Usual use case is to disable tracing for requests so I would do something like

```
message UnsampleSpan {
   Request request
}

message Request {
   string url_pattern // url regex
   string header_key_pattern // header key regex
   string header_value_pattern
}
```

That way we can have different treatment for different objects, e.g. for messages we could have

```
message SQLStatement {
   string query_pattern
}
```

I prefer `sample/unsample` instead of `ignore` as the functionatility collides with sampling. E.g. if ignoring, do you still propagate sampling decision?

**Comment by jcchavezs** [comment]
Shall we close this @mohit-a21 ?

**Pull request closed without merging**